apps/settings/js/panels/sound/tone_manager.js

```diff
@@ -97,20 +97,26 @@ define(function(require) {
         oldRingtoneName = tone.desc.textContent;
       }
       tone.desc.setAttribute('data-l10n-id', 'saving-tone');
+
       var promise;
-      var self = this;
-      // If we got a locked ringtone, we have to unlock it first
       if (result.blob.type.split('/')[1] === ForwardLock.mimeSubtype) {
-        ForwardLock.unlockBlob(secret, result.blob, function(unlocked) {
-          promise = self._isPlayableTone(unlocked);
+        // If we got a locked ringtone, we have to unlock it first.
+        promise = new Promise((resolve, reject) => {
+          ForwardLock.unlockBlob(secret, result.blob, (unlocked) => {
+            resolve(unlocked);
+          });
         });
-      } else {  // Otherwise we can just use the blob directly.
-        promise = self._isPlayableTone(result.blob);
+      } else {
+        // Otherwise we can just use the blob directly.
+        promise = Promise.resolve(result.blob);
       }
-      promise.then(function(isPlayable) {
+
+      promise.then((blob) => {
+        return [blob, this._isPlayableTone(blob)];
+      }).then(([blob, isPlayable]) => {
```

Comment:
Isn't isPlayable an unresolved Promise object at this point? If you return an array of two values, does the Promise system actually check to see if any of those values are promises and wait for them to resolve?

If so, my mind is blown, and please add comments explaining what is happening here, because I this seems complex enough that it will be hard to maintain.

Is there some other way you could structure this code to make it clearer?